### PR TITLE
feat: revamp user storage encryption process

### PR DESCRIPTION
**Pull request by mathieuartu**
## Explanation

This PR adds new steps when both `UserStorageController` and SDK performs:
- `getUserStorageAllFeatureEntries`
- `getUserStorage`

This steps looks for fetched entries' salts and, if random salts are found, re-encrypts these entries with a constant salt and uploads them back to user storage.

This PR also removes the salt randomness when generating the keys, by adding a new shared salt.

This is done to prevent performance issues when decrypting multiple entries that have different salts in applications using `UserStorageController` / SDK.

### `@metamask/profile-sync-controller`

- **CHANGED**: Stop using a random salt when generating scrypt keys and use a shared one
- **ADDED**: Re-encrypt data fetched by `getUserStorageAllFeatureEntries` and `getUserStorage` with the shared salt if fetched entries were encrypted with random salts

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate
- [x] I've prepared draft pull requests for clients and consumer packages to resolve any breaking changes

### CODE REVIEW

File: packages/profile-sync-controller/src/shared/encryption/encryption.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/shared/encryption/encryption.ts:239]: Can you validate both Scrypt KDFs work with an empty salt size?

Comment (Prithpal-Sooriya) [reply packages/profile-sync-controller/src/shared/encryption/encryption.ts:239]: FUTURE - This is a missing test on mobile, but I would like us to add a test to assert that both scrypt functions generate the exact same encrypted data (to avoid the issue we had early on)

File: packages/profile-sync-controller/src/controllers/user-storage/services.test.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/services.test.ts:122]: This seems a bit magical to someone who doesn't know the underlying code (why does it return 'data1'?)

Can we make this a constant?
```ts
const DECRYPED_DATA = 'data1'
const INITIAL_ENCRYPTED_DATA = {
  HashedKey: 'entry1',
  Data: '{"v":"1","t":"scrypt","d":"HIu+WgFBCtKo6rEGy0R8h8t/JgXhzC2a3AF6epahGY2h6GibXDKxSBf6ppxM099Gmg==","o":{"N":131072,"r":8,"p":1,"dkLen":16},"saltLen":16}',
};
```

File: packages/profile-sync-controller/src/controllers/user-storage/services.test.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/services.test.ts:180]: what's with the assertion? Can we avoid the `as string`?

File: packages/profile-sync-controller/src/controllers/user-storage/services.test.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/services.test.ts:203]: FUTURE - this tests have a lot of repetition, it would be nice to clean up these up so the tests clearly show us what we are testing instead of needing to read all this setup/acting.

File: packages/profile-sync-controller/src/controllers/user-storage/services.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/services.ts:98]: Hmmm I wonder if we can instead use the `saltLen` field instead of getting the salt and then checking its length?
```
if(previouslyEncryptedData.Data.saltLen) {
  // re-encrypt with saltLen 0
}
```

Comment (Prithpal-Sooriya) [reply packages/profile-sync-controller/src/controllers/user-storage/services.ts:98]: Oooh I realised the type might be unknown, and we don't have a nice way (yet) to parse and validate the encrypted data shape....

File: packages/profile-sync-controller/src/shared/encryption/cache.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/shared/encryption/cache.ts:59]: Lets reword the JSDOC.

Discuss - should we also rename the function?

File: packages/profile-sync-controller/src/shared/encryption/encryption.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/shared/encryption/encryption.ts:179]: Optional Perf - can we create a separate method called getSaltLength()?

This method does not need to perform the `base64ToByteArray` conversion, and data slicing. Instead it can rely and return the `saltLen`.

Comment (Prithpal-Sooriya) [reply packages/profile-sync-controller/src/shared/encryption/encryption.ts:179]: Main reason I'm bringing this up is that Mobile/React Native has burned us. Because the Hermes engine does not have JIT, performing JS based operations repeatedly can hurt.

(also why I am worried about the AES-GCM JS implementation, this may need to be swapped with a native solution if we find perf issues)

Comment (Prithpal-Sooriya) [reply packages/profile-sync-controller/src/shared/encryption/encryption.ts:179]: This is optional - so I don't want it blocking this PR :)

File: packages/profile-sync-controller/src/shared/encryption/utils.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/shared/encryption/utils.ts:29]: Could this be simplified?
```ts
const strSet = new Set(arrays.map(arr => arr.toString()))
return strSet.size === arrays.length
```

But yeah your func if more performant (since it bails earlier)

File: packages/profile-sync-controller/src/shared/encryption/encryption.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/shared/encryption/encryption.ts:197]: (you can close this comment, not urgent) Are we sure we want to throw?
Technically we should be safe since the places that use this already to a decrypt first.

Comment (Prithpal-Sooriya) [reply packages/profile-sync-controller/src/shared/encryption/encryption.ts:197]: Could we add a test case I guess?

File: packages/profile-sync-controller/src/shared/encryption/encryption.ts
Comment (mathieuartu) [reply packages/profile-sync-controller/src/shared/encryption/encryption.ts:239]: > Can you validate both Scrypt KDFs work with an empty salt size?

I CAN! 😄

File: packages/profile-sync-controller/src/controllers/user-storage/services.test.ts
Comment (mathieuartu) [reply packages/profile-sync-controller/src/controllers/user-storage/services.test.ts:181]: That's because `requestBody.data` is of type `any`, and `Object.entries` coerces it as `unknown`. Double punishment

File: packages/profile-sync-controller/src/shared/encryption/encryption.ts
Comment (mirceanis) [reply packages/profile-sync-controller/src/shared/encryption/encryption.ts:239]: I'd argue that if we want to avoid running into multiple salts we can hardcode a salt here instead of fiddling with the size.
I'm also concerned about the behavior of different libraries when given empty salts

File: packages/profile-sync-controller/src/shared/encryption/utils.ts
Comment (mathieuartu) [reply packages/profile-sync-controller/src/shared/encryption/utils.ts:42]: We also need them for SDK tests, in order to avoid duplication and / or import spillage, I've put them in the shared folder. But you're right, this creates a circular dep. Should we assume this would get tree-shaken?

Anyway, I've moved this util function directly inside `encryption.ts` and added the tests in `encryption.test.ts`!

Comment (Prithpal-Sooriya) [reply packages/profile-sync-controller/src/shared/encryption/utils.ts]: Haha, I would assume tree shaking... But we are using this in extension and mobile lol.

You could validate by making a preview build, and testing on extension and mobile to ensure builds are okay.

File: packages/profile-sync-controller/src/controllers/user-storage/services.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/services.ts:99]: Nice new check!

File: packages/profile-sync-controller/src/controllers/user-storage/services.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/controllers/user-storage/services.ts:280]: ye, duplication is pain. But we can clean up and get a nice abstraction later!

File: packages/profile-sync-controller/src/sdk/user-storage.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/sdk/user-storage.ts:204]: Future - I don't like how we've shoved all the user storage logic in a class, this bloats the SDK class and makes it harder to test things individually.

Similar to the authentication SDK, we should split out `services` from the user storage SDK.

File: packages/profile-sync-controller/src/shared/encryption/encryption.test.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/shared/encryption/encryption.test.ts:101]: Although this is only used for tests, this is fine. At least no circular deps.

File: packages/profile-sync-controller/src/shared/encryption/encryption.ts
Comment (Prithpal-Sooriya) [reply packages/profile-sync-controller/src/shared/encryption/encryption.ts:179]: OK we cannot do the `getSaltLen` approach as we have now opted for a shared salt.

I am closing this discussion, but keep in mind the AES-GCM JS performance on mobile. This has not been thoroughly tested.

File: packages/profile-sync-controller/src/shared/encryption/encryption.ts
Comment (Prithpal-Sooriya) [inline packages/profile-sync-controller/src/shared/encryption/encryption.ts:200]: This is fine... although it is only used for testing purposes.

**Review by Prithpal-Sooriya — DISMISSED** [review]
LGTM.

Test this on mobile before merging. I want to ensure that both JS and Native Scrypt will generate same key / same encrypted payload.

File: packages/profile-sync-controller/src/sdk/user-storage.ts
Comment (mathieuartu) [reply packages/profile-sync-controller/src/sdk/user-storage.ts:204]: @Prithpal-Sooriya Agreed! Added that to my to-do list!

**Review by mirceanis — CHANGES_REQUESTED** [review]
Overall looks good.

There's a spot where the comparison is still done by salt.length instead of comparing to the shared salt.

Once that's fixed it should be good to go.

**Review by mirceanis — DISMISSED** [review]
looks good

File: packages/profile-sync-controller/src/sdk/user-storage.ts
Comment (mirceanis) [reply packages/profile-sync-controller/src/sdk/user-storage.ts]: I wonder if it's difficult to cover this in a test too

**Review by mirceanis — CHANGES_REQUESTED** [review]
I disagree with the approach to delete the data that fails to decrypt.
Such failures can happen for multiple unforeseen reasons so any such destructive operation should be guarded much more thoroughly.

If we intend to avoid spending time in KDF on seemingly corrupted entries perhaps an alternative approach can be considered: marking the corrupted entries locally so that decryption attempts are stopped before KDF is triggered.

File: packages/profile-sync-controller/src/controllers/user-storage/services.ts
Comment (mirceanis) [inline packages/profile-sync-controller/src/controllers/user-storage/services.ts:176]: Let's address this in a separate issue.

**Pull request merged**
